### ISSUE TITLE
Type Checker getting confused with constraints + ICE

### ISSUE THREAD

**Issue by CryZe** [rust-lang/rust#33904]
I've written the same code in 2 different crates. One compiles perfectly fine, the other doesn't. Apparently the constraints on the function merge(!!) with the constraints of the other function and the type checker is getting really confused and errors out. If you look at the code, the constrained type Q does not come in contact with convex_hull::create in any way, yet it influences it. Removing Q from the function lets convex_hull::create compile just fine.

Also the compiler is complaining about a type called convex_hull::Vector3, which doesn't exist, it's called Vec3. Vec3 is actually a reexport of Vector3 as Vec3, which then got reexported again (and I think again). So maybe somewhere the type checker got confused by that and now thinks it needs to look for a type called Vector3. Not sure if that's related to the problem at all though, just seems like maybe another small bug.

![http://i.imgur.com/DdOIt9t.png](http://i.imgur.com/DdOIt9t.png)

**Comment by eddyb** [comment]
Looks like this is #30472.

**Comment by CryZe** [comment]
It's not, it also gets confused if I remove the HRTB and add the lifetime
to the function's generics.

Am Freitag, 27. Mai 2016 schrieb Eduard-Mihai Burtescu :

**Comment by eddyb** [comment]
@CryZe With associated types under HRTB, the compiler cannot tell that two identical-looking bounds are the same, or that they imply each-other. This has been a problem for as long as HRTB has existed.

**Comment by CryZe** [comment]
![http://puu.sh/p72qF.png](http://puu.sh/p72qF.png)

As you can see here, I just removed all HRTB and it still doesn't compile.

**Comment by CryZe** [comment]
This is just getting weirder and weirder:

![http://puu.sh/p7342.png](http://puu.sh/p7342.png)

**Comment by eddyb** [comment]
The Rust compiler **never** uses type parameter names from callees in error messages in callers.
It's type-checking `concave_hull::create`, which has a `Q` parameter, so that's what it will say.

**Comment by eddyb** [comment]
@CryZe Could you reduce to one file that we can reasonably work with? (The ICE is likely #33852 AFAICT)
I don't see any reason having two crates is actually relevant here, other than potential confusion.

**Comment by CryZe** [comment]
The ICE seems to be cause by the Vec not being able to type infer its type from the constraint. Explicitly giving it a type makes it work.

**Comment by CryZe** [comment]
Here's the reduced code: https://is.gd/OOg72m

ICE happens on Beta and Nightly. Type Checker problem happens on all.

```rust
pub fn problem_creator<'a, P>(points: &'a P) -> Option<usize>
    where &'a P: IntoIterator<Item=&'a f64>
{
    unimplemented!()
}

pub fn ok() {
    let points = Vec::new();
    let mut hull = problem_creator(&points).unwrap();
}

pub fn ice<'a, Q>(_: &'a Q)
    where &'a Q: IntoIterator<Item=&'a f64>
{
    let points = Vec::new();
    let mut hull = problem_creator(&points).unwrap();
}

pub fn broken_typechecking<'a, Q>(_: &'a Q)
    where &'a Q: IntoIterator<Item=&'a f64>
{
    let points: Vec<f64> = Vec::new();
    let mut hull = problem_creator(&points).unwrap();
}

fn main() {}
```

**Comment by eddyb** [comment]
Ahh, I see what happens: `problem_creator` expects `&$0: IntoIterator` and the bound is actually satisfied by `&Q: IntoIterator`, which results in `$0 == Q` *even though* `points` has type `Vec<f64>`.
However, obligations like `&$0: IntoIterator` get resolved before arguments are checked, so that's how this arises.
I suspect #33852 fixes the ICE, but I'm not sure. cc @nikomatsakis @arielb1

**Comment by eddyb** [comment]
@CryZe As a workaround, `problem_creator::<Vec<_>>(&points)` seems to work for me, could you try that (explicitly specifying type parameters) in your own project?

**Comment by CryZe** [comment]
Yeah, seems to work. Thanks for your help :)

**Comment by arielb1** [comment]
@eddyb 

That's the standard "inference guessing" issue (#30767).

**Comment by eddyb** [comment]
@arielb1 Is the ICE fixed?

**Comment by arielb1** [comment]
@eddyb 

it looks like the `Vec` one, but that didn't reach nightly yet.
